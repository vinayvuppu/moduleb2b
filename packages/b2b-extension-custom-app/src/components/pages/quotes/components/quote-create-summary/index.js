export { default } from './quote-create-summary';
