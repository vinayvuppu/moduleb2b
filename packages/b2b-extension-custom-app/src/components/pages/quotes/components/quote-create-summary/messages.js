import { defineMessages } from 'react-intl';

export default defineMessages({
  itemsLabel: {
    id: 'QuoteSummary.itemsLabel',
    description: 'items label when items are selected',
    defaultMessage: '{items} items',
  },
  quoteSummaryTittle: {
    id: 'QuoteSummary.quoteSummaryTittle',
    description: 'title for the cart summary panel',
    defaultMessage: 'in your quote',
  },
  quantityLabel: {
    id: 'QuoteSummary.quantityLabel',
    description: 'label for the line item quantity',
    defaultMessage: 'Qty: {quantity}',
  },
  subtotalLabel: {
    id: 'QuoteSummary.subtotalLabel',
    description: 'label for the cart subtotal',
    defaultMessage: 'Subtotal',
  },
  totalLabel: {
    id: 'QuoteSummary.totalLabel',
    description: 'label for the cart total',
    defaultMessage: 'Total (gross)',
  },
  emptyQuoteLabel: {
    id: 'QuoteSummary.emptyQuoteLabel',
    description: 'label for the empty shopping cart message',
    defaultMessage: 'Your quote is empty',
  },
  removeLineItem: {
    id: 'QuoteSummary.removeLineItem',
    defaultMessage: 'Remove item',
  },
  changeQuantity: {
    id: 'QuoteSummary.changeQuantity',
    defaultMessage: 'Change quantity',
  },
});
