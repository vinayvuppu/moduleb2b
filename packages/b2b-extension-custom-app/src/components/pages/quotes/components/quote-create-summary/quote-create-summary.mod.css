.separator {
  display: block;
  height: 1px;
  border: 0;
  border-top: 1px solid var(--color-neutral);
  padding: 0;
}
.was {
  text-decoration: line-through;
}
