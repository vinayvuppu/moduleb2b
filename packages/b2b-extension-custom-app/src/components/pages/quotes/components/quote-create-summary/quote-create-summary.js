import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage, useIntl } from 'react-intl';

import {
  Card,
  CartIcon,
  Spacings,
  Text,
  IconButton,
  CloseBoldIcon,
  EditIcon,
} from '@commercetools-frontend/ui-kit';
import { formatMoney } from '@commercetools-local/utils/formats/money';
import { transformLocalizedFieldToString } from '@commercetools-local/utils/graphql';
import * as priceUtils from '@commercetools-local/utils/prices';
import Localize from '@commercetools-local/core/components/localize';

import { selectTotalGrossPrice } from '../../../my-company/utils/cart-selectors';
import * as orderPriceUtils from '../../../my-company/utils/order-prices';
import ChangeQuantityModal from '../../../my-company/components/order-create-cart-summary/change-quantity-modal';

import styles from './quote-create-summary.mod.css';
import messages from './messages';

export const getVariantAttributes = attributes =>
  attributes.filter(
    attr => attr.attributeDefinition.attributeConstraint === 'CombinationUnique'
  );

const SummaryItem = props => (
  <Spacings.Inline justifyContent="space-between">
    {props.children}
  </Spacings.Inline>
);

SummaryItem.displayName = 'SummaryItem';
SummaryItem.propTypes = {
  children: PropTypes.node.isRequired,
};

export const QuoteCreateSummary = props => {
  const intl = useIntl();
  const [changeQuantityOptions, setChangeQuantityOptions] = useState({
    isChangeQuantityModalOpen: false,
    changeQuantityLineItem: undefined,
  });

  const handleSelectLineItemToChangeQuantity = lineItem =>
    setChangeQuantityOptions({
      isChangeQuantityModalOpen: true,
      changeQuantityLineItem: lineItem,
    });

  const handleCloseModal = () =>
    setChangeQuantityOptions({
      isChangeQuantityModalOpen: false,
      changeQuantityLineItem: undefined,
    });

  const handleChangeQuantity = async quantity => {
    await props.onChangeLineItemQuantity({
      lineItemId: changeQuantityOptions.changeQuantityLineItem.id,
      quantity: +quantity,
    });

    handleCloseModal();
  };

  const itemsNumber = props.quote?.lineItems.length;
  const lineItems = props.quote?.lineItems || [];

  const totalGross = selectTotalGrossPrice(props.quote);
  const subtotal = orderPriceUtils.getNetPriceWithoutShipping(props.quote);

  return (
    <Card>
      <Spacings.Stack>
        <Spacings.Inline alignItems="center" scale="xs">
          <CartIcon size="medium" />
          <Text.Detail
            isBold={true}
            tone="information"
            intlMessage={{
              ...messages.itemsLabel,
              values: { items: itemsNumber },
            }}
          />
          <Text.Detail isBold={true}>
            <FormattedMessage {...messages.quoteSummaryTittle} />
          </Text.Detail>
        </Spacings.Inline>
        <div className={styles.separator} />
        {itemsNumber > 0 ? (
          lineItems.map(lineItem => (
            <SummaryItem key={lineItem.id}>
              <Spacings.Inline>
                <IconButton
                  size="small"
                  data-testid={`remove-line-item-${lineItem.id}`}
                  icon={<CloseBoldIcon />}
                  label={intl.formatMessage(messages.removeLineItem)}
                  onClick={() => props.onRemoveLineItem(lineItem.id)}
                />
                <Spacings.Stack>
                  <Text.Detail isBold={true}>
                    {typeof lineItem.name === 'string' ? (
                      lineItem.name
                    ) : (
                      <Localize
                        object={{
                          name: transformLocalizedFieldToString(
                            lineItem.nameAllLocales
                          ),
                        }}
                        objectKey="name"
                      />
                    )}
                  </Text.Detail>
                  <Text.Detail tone="secondary">
                    {lineItem.variant.sku}
                  </Text.Detail>
                </Spacings.Stack>
              </Spacings.Inline>
              <Spacings.Stack alignItems="flexEnd">
                <Text.Detail tone="secondary">
                  {formatMoney(
                    lineItem.price
                      ? priceUtils.getNetUnitPrice({
                          lineItem,
                          shouldRoundAmount: true,
                        })
                      : lineItem.money,
                    intl
                  )}
                </Text.Detail>
                <Spacings.Inline>
                  <IconButton
                    size="small"
                    icon={<EditIcon />}
                    label={intl.formatMessage(messages.changeQuantity)}
                    onClick={() =>
                      handleSelectLineItemToChangeQuantity(lineItem)
                    }
                  />
                  <Text.Detail isBold={true}>
                    <FormattedMessage
                      {...messages.quantityLabel}
                      values={{ quantity: lineItem.quantity }}
                    />
                  </Text.Detail>
                </Spacings.Inline>
                <Text.Detail isBold={true}>
                  {formatMoney(lineItem.totalPrice, intl)}
                </Text.Detail>
              </Spacings.Stack>
            </SummaryItem>
          ))
        ) : (
          <Text.Detail tone="secondary">
            <FormattedMessage {...messages.emptyQuoteLabel} />
          </Text.Detail>
        )}
        <div className={styles.separator} />
        <SummaryItem>
          <Text.Detail isBold={true}>
            <FormattedMessage {...messages.subtotalLabel} />
          </Text.Detail>
          <Text.Detail isBold={true}>
            {formatMoney(
              {
                ...subtotal,
                centAmount: subtotal.centAmount,
              },
              intl
            )}
          </Text.Detail>
        </SummaryItem>

        <div className={styles.separator} />
        <SummaryItem>
          <Text.Body isBold={true}>
            <FormattedMessage {...messages.totalLabel} />
          </Text.Body>
          <Text.Body isBold={true}>{formatMoney(totalGross, intl)}</Text.Body>
        </SummaryItem>
      </Spacings.Stack>
      {changeQuantityOptions.isChangeQuantityModalOpen && (
        <ChangeQuantityModal
          quantity={changeQuantityOptions.changeQuantityLineItem?.quantity}
          productName={changeQuantityOptions.changeQuantityLineItem?.name}
          isOpen={changeQuantityOptions.isChangeQuantityModalOpen}
          handleOnClose={handleCloseModal}
          handleOnSave={handleChangeQuantity}
        />
      )}
    </Card>
  );
};

QuoteCreateSummary.displayName = 'QuoteCreateSummary';
QuoteCreateSummary.propTypes = {
  quote: PropTypes.object.isRequired,
  onRemoveLineItem: PropTypes.func.isRequired,
  onChangeLineItemQuantity: PropTypes.func.isRequired,
};

export default QuoteCreateSummary;
