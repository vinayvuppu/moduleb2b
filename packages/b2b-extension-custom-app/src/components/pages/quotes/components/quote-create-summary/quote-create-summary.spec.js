import React from 'react';
import { shallow } from 'enzyme';

import { QuoteCreateSummary } from './quote-create-summary';

jest.mock('react-intl', () => {
  const actual = jest.requireActual('react-intl');
  return {
    ...actual,
    useIntl: jest.fn(() => ({
      formatNumber: jest.fn(({ centAmount }) => ({ centAmount })),
      formatMessage: jest.fn(() => 'formatted message'),
    })),
  };
});

const createTestProps = props => ({
  quote: {
    id: 'quote-id',
    company: {
      id: 'company-id',
      name: 'company name',
    },
    employeeEmail: 'employee@company.com',
    totalPrice: {
      currencyCode: 'USD',
      centAmount: 100,
    },
    lineItems: [
      {
        id: 'line-item-id',
        productId: 'product-id',
        nameAllLocales: [
          {
            locale: 'en',
            value: 'Top Jucca black',
          },
        ],
        variant: {
          sku: 'M0E20000000EC8N',
          images: [
            {
              url: 'url-1',
            },
          ],
        },
        price: {
          id: 'price-id-1',
          value: {
            currencyCode: 'USD',
            centAmount: 100,
          },
        },
        quantity: 1,
        totalPrice: {
          currencyCode: 'USD',
          centAmount: 100,
        },
      },
    ],
  },
  onRemoveLineItem: jest.fn(),
  onChangeLineItemQuantity: jest.fn(),
  ...props,
});

describe('render', () => {
  let props;
  let wrapper;

  describe('when no line items', () => {
    beforeEach(() => {
      props = createTestProps({ lineItems: [] });
      wrapper = shallow(<QuoteCreateSummary {...props} />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('when line items', () => {
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<QuoteCreateSummary {...props} />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});

describe('callbacks', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<QuoteCreateSummary {...props} />);
  });

  describe('when remove line item', () => {
    beforeEach(() => {
      wrapper
        .find({ 'data-testid': 'remove-line-item-line-item-id' })
        .simulate('click');
    });

    it('should call to onRemoveLineItem', () => {
      expect(props.onRemoveLineItem).toHaveBeenCalledWith('line-item-id');
    });
  });
});
