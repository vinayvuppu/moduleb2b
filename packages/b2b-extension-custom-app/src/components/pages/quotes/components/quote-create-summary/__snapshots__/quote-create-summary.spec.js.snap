// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render when line items should match snapshot 1`] = `
<Card
  theme="light"
  type="raised"
>
  <Stack
    alignItems="stretch"
    scale="s"
  >
    <Inline
      alignItems="center"
      justifyContent="flex-start"
      scale="xs"
    >
      <CartIcon
        size="medium"
      />
      <TextDetail
        intlMessage={
          Object {
            "defaultMessage": "{items} items",
            "description": "items label when items are selected",
            "id": "QuoteSummary.itemsLabel",
            "values": Object {
              "items": 1,
            },
          }
        }
        isBold={true}
        tone="information"
      />
      <TextDetail
        isBold={true}
      >
        <FormattedMessage
          defaultMessage="in your quote"
          description="title for the cart summary panel"
          id="QuoteSummary.quoteSummaryTittle"
          values={Object {}}
        />
      </TextDetail>
    </Inline>
    <div
      className="separator"
    />
    <SummaryItem
      key="line-item-id"
    >
      <Inline
        alignItems="flex-start"
        justifyContent="flex-start"
        scale="s"
      >
        <IconButton
          data-testid="remove-line-item-line-item-id"
          icon={<CloseBoldIcon />}
          isToggleButton={false}
          label="formatted message"
          onClick={[Function]}
          shape="round"
          size="small"
          theme="default"
          type="button"
        />
        <Stack
          alignItems="stretch"
          scale="s"
        >
          <TextDetail
            isBold={true}
          >
            <withApplicationContext(Localize)
              object={
                Object {
                  "name": Object {
                    "en": "Top Jucca black",
                  },
                }
              }
              objectKey="name"
            />
          </TextDetail>
          <TextDetail
            tone="secondary"
          >
            M0E20000000EC8N
          </TextDetail>
        </Stack>
      </Inline>
      <Stack
        alignItems="flexEnd"
        scale="s"
      >
        <TextDetail
          tone="secondary"
        >
          <Component />
        </TextDetail>
        <Inline
          alignItems="flex-start"
          justifyContent="flex-start"
          scale="s"
        >
          <IconButton
            icon={<EditIcon />}
            isToggleButton={false}
            label="formatted message"
            onClick={[Function]}
            shape="round"
            size="small"
            theme="default"
            type="button"
          />
          <TextDetail
            isBold={true}
          >
            <FormattedMessage
              defaultMessage="Qty: {quantity}"
              description="label for the line item quantity"
              id="QuoteSummary.quantityLabel"
              values={
                Object {
                  "quantity": 1,
                }
              }
            />
          </TextDetail>
        </Inline>
        <TextDetail
          isBold={true}
        >
          <Component />
        </TextDetail>
      </Stack>
    </SummaryItem>
    <div
      className="separator"
    />
    <SummaryItem>
      <TextDetail
        isBold={true}
      >
        <FormattedMessage
          defaultMessage="Subtotal"
          description="label for the cart subtotal"
          id="QuoteSummary.subtotalLabel"
          values={Object {}}
        />
      </TextDetail>
      <TextDetail
        isBold={true}
      >
        <Component />
      </TextDetail>
    </SummaryItem>
    <div
      className="separator"
    />
    <SummaryItem>
      <TextBody
        isBold={true}
      >
        <FormattedMessage
          defaultMessage="Total (gross)"
          description="label for the cart total"
          id="QuoteSummary.totalLabel"
          values={Object {}}
        />
      </TextBody>
      <TextBody
        isBold={true}
      >
        <Component />
      </TextBody>
    </SummaryItem>
  </Stack>
</Card>
`;

exports[`render when no line items should match snapshot 1`] = `
<Card
  theme="light"
  type="raised"
>
  <Stack
    alignItems="stretch"
    scale="s"
  >
    <Inline
      alignItems="center"
      justifyContent="flex-start"
      scale="xs"
    >
      <CartIcon
        size="medium"
      />
      <TextDetail
        intlMessage={
          Object {
            "defaultMessage": "{items} items",
            "description": "items label when items are selected",
            "id": "QuoteSummary.itemsLabel",
            "values": Object {
              "items": 1,
            },
          }
        }
        isBold={true}
        tone="information"
      />
      <TextDetail
        isBold={true}
      >
        <FormattedMessage
          defaultMessage="in your quote"
          description="title for the cart summary panel"
          id="QuoteSummary.quoteSummaryTittle"
          values={Object {}}
        />
      </TextDetail>
    </Inline>
    <div
      className="separator"
    />
    <SummaryItem
      key="line-item-id"
    >
      <Inline
        alignItems="flex-start"
        justifyContent="flex-start"
        scale="s"
      >
        <IconButton
          data-testid="remove-line-item-line-item-id"
          icon={<CloseBoldIcon />}
          isToggleButton={false}
          label="formatted message"
          onClick={[Function]}
          shape="round"
          size="small"
          theme="default"
          type="button"
        />
        <Stack
          alignItems="stretch"
          scale="s"
        >
          <TextDetail
            isBold={true}
          >
            <withApplicationContext(Localize)
              object={
                Object {
                  "name": Object {
                    "en": "Top Jucca black",
                  },
                }
              }
              objectKey="name"
            />
          </TextDetail>
          <TextDetail
            tone="secondary"
          >
            M0E20000000EC8N
          </TextDetail>
        </Stack>
      </Inline>
      <Stack
        alignItems="flexEnd"
        scale="s"
      >
        <TextDetail
          tone="secondary"
        >
          <Component />
        </TextDetail>
        <Inline
          alignItems="flex-start"
          justifyContent="flex-start"
          scale="s"
        >
          <IconButton
            icon={<EditIcon />}
            isToggleButton={false}
            label="formatted message"
            onClick={[Function]}
            shape="round"
            size="small"
            theme="default"
            type="button"
          />
          <TextDetail
            isBold={true}
          >
            <FormattedMessage
              defaultMessage="Qty: {quantity}"
              description="label for the line item quantity"
              id="QuoteSummary.quantityLabel"
              values={
                Object {
                  "quantity": 1,
                }
              }
            />
          </TextDetail>
        </Inline>
        <TextDetail
          isBold={true}
        >
          <Component />
        </TextDetail>
      </Stack>
    </SummaryItem>
    <div
      className="separator"
    />
    <SummaryItem>
      <TextDetail
        isBold={true}
      >
        <FormattedMessage
          defaultMessage="Subtotal"
          description="label for the cart subtotal"
          id="QuoteSummary.subtotalLabel"
          values={Object {}}
        />
      </TextDetail>
      <TextDetail
        isBold={true}
      >
        <Component />
      </TextDetail>
    </SummaryItem>
    <div
      className="separator"
    />
    <SummaryItem>
      <TextBody
        isBold={true}
      >
        <FormattedMessage
          defaultMessage="Total (gross)"
          description="label for the cart total"
          id="QuoteSummary.totalLabel"
          values={Object {}}
        />
      </TextBody>
      <TextBody
        isBold={true}
      >
        <Component />
      </TextBody>
    </SummaryItem>
  </Stack>
</Card>
`;
